src/protocols/rdp/rdp.c

```diff
@@ -0,0 +1,931 @@
+ * parent keymap, if any. These mappings are stored within the guac_rdp_client
+ * structure associated with the given guac_client for future use in
+ * translating keysyms to the scancodes required by RDP key events.
+ *
+ * @param client
+ *     The guac_client whose associated guac_rdp_client should be initialized
+ *     with the keysym/scancode mapping defined in the given keymap.
+ *
+ * @param keymap
+ *     The keymap to use to populate the given client's keysym/scancode
+ *     mapping.
+ */
+static void __guac_rdp_client_load_keymap(guac_client* client,
+        const guac_rdp_keymap* keymap) {
+
+    guac_rdp_client* rdp_client =
+        (guac_rdp_client*) client->data;
+
+    /* Get mapping */
+    const guac_rdp_keysym_desc* mapping = keymap->mapping;
+
+    /* If parent exists, load parent first */
+    if (keymap->parent != NULL)
+        __guac_rdp_client_load_keymap(client, keymap->parent);
+
+    /* Log load */
+    guac_client_log(client, GUAC_LOG_INFO, "Loading keymap \"%s\"", keymap->name);
+
+    /* Load mapping into keymap */
+    while (mapping->keysym != 0) {
+
+        /* Copy mapping */
+        GUAC_RDP_KEYSYM_LOOKUP(rdp_client->keymap, mapping->keysym) =
+            *mapping;
+
+        /* Next keysym */
+        mapping++;
+
+    }
+
+}
+
+/**
+ * Waits for messages from the RDP server for the given number of microseconds.
+ *
+ * @param client
+ *     The client associated with the current RDP session.
+ *
+ * @param timeout_usecs
+ *     The maximum amount of time to wait, in microseconds.
+ *
+ * @return
+ *     A positive value if messages are ready, zero if the specified timeout
+ *     period elapsed, or a negative value if an error occurs.
+ */
+static int rdp_guac_client_wait_for_messages(guac_client* client,
+        int timeout_usecs) {
+
+    guac_rdp_client* rdp_client = (guac_rdp_client*) client->data;
+    freerdp* rdp_inst = rdp_client->rdp_inst;
+    rdpChannels* channels = rdp_inst->context->channels;
+
+    int result;
+    int index;
+    int max_fd, fd;
+    void* read_fds[32];
+    void* write_fds[32];
+    int read_count = 0;
+    int write_count = 0;
+    fd_set rfds, wfds;
+
+    struct timeval timeout = {
+        .tv_sec  = 0,
+        .tv_usec = timeout_usecs
+    };
+
+    /* Get RDP fds */
+    if (!freerdp_get_fds(rdp_inst, read_fds, &read_count, write_fds, &write_count)) {
+        guac_client_abort(client, GUAC_PROTOCOL_STATUS_SERVER_ERROR, "Unable to read RDP file descriptors.");
+        return -1;
+    }
+
+    /* Get channel fds */
+    if (!freerdp_channels_get_fds(channels, rdp_inst, read_fds, &read_count, write_fds,
+                &write_count)) {
+        guac_client_abort(client, GUAC_PROTOCOL_STATUS_SERVER_ERROR, "Unable to read RDP channel file descriptors.");
+        return -1;
+    }
+
+    /* Construct read fd_set */
+    max_fd = 0;
+    FD_ZERO(&rfds);
+    for (index = 0; index < read_count; index++) {
+        fd = (int)(long) (read_fds[index]);
```

Comment:
Well...

`read_fds` and `write_fds` in this breed of functions from FreeRDP are both `void**`. Typecasting those directly to `int` can't work:

```
test.c:9:14: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
```

Typecasting from `void**` to `long` to `int` will work, however.